### PR TITLE
Consolidated EIP-4844

### PR DESCRIPTION
**Pull request by jflo**
Implements EIP-4844.

- introduces a Hardfork class to the protocol schedule system
- new Engine APIs required for CL to work on 4844
- refactors Engine API tests to reuse OR override prior versions
- new DataGas type for tracking block cost for 4844 data
- new VersionedHash type to reflect that a versioned hash is not quite a pure sha256
- incorporates wrapped jc-kzg library for KZG point evaluations
- New transaction type, and domain objects for constituent parts to represent the Blobs, KZGCommitments, and Proofs used for 4844
- RLP encoders and decoders to support new transaction type
- gas pricing calculators for the new type of gas
- plugin-api version was changed

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/AbstractEngineNewPayload.java [hunk 1]
```diff
@@ -97,13 +104,32 @@ public JsonRpcResponse syncResponse(final JsonRpcRequestContext requestContext)
...
         .setMessage("blockparam: {}")
         .addArgument(() -> Json.encodePrettily(blockParam))
         .log();
 
+    /*
+    ValidationResult<JsonRpcError> forkValidationResult = validateForkSupported(reqId, blockParam);
+    if (!forkValidationResult.isValid()) {
+      return new JsonRpcErrorResponse(reqId, forkValidationResult.getInvalidReason());
+    }
+    */
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/AbstractEngineNewPayload.java:131]: is this left in deliberately?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java [hunk 2]
```diff
@@ -249,4 +281,109 @@ private ValidationResult<TransactionInvalidReason> validateTransactionSignature(
...
+    return ValidationResult.valid();
+  }
+
+  /*
+  private VersionedHash hashCommitment(final Bytes32 commitment) {
+    return new VersionedHash(
+        VersionedHash.SHA256_VERSION_ID, Sha256Hash.hash(commitment));
+  }
+
+   */
```
Comment (macfarla) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionValidator.java:376]: deliberately left in here?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/feemarket/ExcessDataGasCalculator.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+package org.hyperledger.besu.ethereum.mainnet.feemarket;
+
+import org.hyperledger.besu.datatypes.DataGas;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSpec;
+
+/** Calculates the excess data gas for a parent block header. */
+public class ExcessDataGasCalculator {
+  /**
+   * public class ExcessDataGasCalculator { /** Calculates the excess data gas for a parent block
```
Comment (macfarla) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/feemarket/ExcessDataGasCalculator.java:24]:
```suggestion
   * Calculates the excess data gas for a parent block
```

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/mainnet/feemarket/ZeroBaseFeeMarketTest.java [hunk 4]
```diff
@@ -147,6 +147,6 @@ public void implementsDataFeedShouldReturnFalse() {
 
   @Test
   public void dataPriceShouldReturnsZero() {
```
Comment (macfarla) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/mainnet/feemarket/ZeroBaseFeeMarketTest.java:149]:
```suggestion
  public void dataPricePerGasShouldReturnZero() {
```

File: ethereum/core/src/test/resources/org/hyperledger/besu/ethereum/core/encoding/BlobDataFixture.bin
Comment (macfarla) [inline ethereum/core/src/test/resources/org/hyperledger/besu/ethereum/core/encoding/BlobDataFixture.bin:1]: lol

**Review by macfarla — APPROVED** [review]
LGTM

**Review by fab-10 — APPROVED** [review]

**Comment by garyschulte** [comment]
congrats @jflo, a monumental merge

**Pull request merged**